controllers/observability/perses.go

```diff
@@ -0,0 +1,227 @@
+	"os"
+	"path/filepath"
+
+	authv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// persesAssetsDir is the directory (inside the repo image) containing Perses CR manifests (yaml).
+// Each YAML should define a single resource: PersesDashboard or PersesDatasource.
+const persesAssetsDir = "assets/dashboards/perses"
+
+// ReconcilePersesResources applies all Perses resources shipped with the operator under assets/perses.
+// It uses server-side apply to create/update resources idempotently without relying on cache reads.
+func (r *Reconciler) ReconcilePersesResources(ctx context.Context) error {
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get working dir: %w", err)
+	}
+
+	d := os.DirFS(wd)
+	// If the directory does not exist, nothing to do.
+	if _, err := fs.Stat(d, persesAssetsDir); err != nil {
+		return nil
+	}
+
+	// Ensure Perses CRDs are present before attempting to apply resources.
+	// Without CRDs, server-side apply will fail on every reconcile loop.
+	if !r.persesCRDsReady(ctx) {
+		return nil
+	}
+
+	// Load and cache assets once
+	var loadErr error
+	r.assetsOnce.Do(func() {
+		r.cachedAssets, loadErr = readPersesAssets(d)
+	})
+	if loadErr != nil {
+		return loadErr
+	}
+
+	// Apply cached assets each reconcile
+	for _, raw := range r.cachedAssets {
+		objMap := deepCopyMap(raw)
+		// enforce namespace
+		metadata, _ := objMap["metadata"].(map[string]any)
+		if metadata == nil {
+			metadata = map[string]any{}
+			objMap["metadata"] = metadata
+		}
+		metadata["namespace"] = r.namespace
+
+		apiVersion, _ := objMap["apiVersion"].(string)
+		kind, _ := objMap["kind"].(string)
+		if apiVersion == "" || kind == "" {
+			return fmt.Errorf("missing apiVersion/kind in cached asset")
+		}
+		u := &unstructured.Unstructured{Object: objMap}
+		u.SetGroupVersionKind(schema.FromAPIVersionAndKind(apiVersion, kind))
+
+		if err := r.Patch(ctx, u, client.Apply, client.FieldOwner("hyperconverged-operator"), client.ForceOwnership); err != nil {
+			return fmt.Errorf("failed to apply %s/%s: %w", u.GetNamespace(), u.GetName(), err)
+		}
+	}
+	return nil
+}
+
+// persesCRDsReady returns true when Perses CRDs exist in the cluster.
+func (r *Reconciler) persesCRDsReady(ctx context.Context) bool {
+	crdGVK := schema.GroupVersionKind{
+		Group:   "apiextensions.k8s.io",
+		Version: "v1",
+		Kind:    "CustomResourceDefinition",
+	}
+	for _, crdName := range []string{
+		"persesdashboards.perses.dev",
+		"persesdatasources.perses.dev",
+	} {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(crdGVK)
+		if err := r.Get(ctx, types.NamespacedName{Name: crdName}, u); err != nil {
+			return false
+		}
+	}
+	return true
```

Comment:
just read into the CRD go type. No need to use unstructured here. Also, please check the existing CRD controller - maybe you can reuse it somehow?

Reply:
We have a utility here https://github.com/kubevirt/hyperconverged-cluster-operator/blob/cae67e758968fda8c1c71efa1509f0f917897465/pkg/util/cluster.go#L228 

Please reuse it: either expose it (rename it to `IsCRDExists`), or add a new utility; e.g.
```go
func IsPersesAvailable(ctx context.Context, cl client.Client) bool {
	return isCRDExists(ctx, cl, "persesdashboards.perses.dev") &&
			isCRDExists(ctx, cl, "persesdatasources.perses.dev") 
}